security-oauth2/src/main/java/io/micronaut/security/oauth2/grants/AbstractClientSecureGrant.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2017-2022 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.security.oauth2.grants;
+
+import io.micronaut.core.annotation.NonNull;
+
+import java.util.Map;
+
+/**
+ * Base class for {@link SecureGrant} implementations.
+ *
+ * @author Álvaro Sánchez-Mariscal
+ */
+public abstract class AbstractClientSecureGrant implements SecureGrant, AsMap {
+
+    private static final String KEY_GRANT_TYPE = "grant_type";
+
+    protected String clientId;
```

Comment:
annotate with nullability annotation

Reply:
Nope. Non-null fields must be initialised, either inline or in a constructor. Moreover, they are protected fields. Their public accessors are what need the `@NonNull` annotation.

Reply:
It seems client id is optional too, despite what its getter indicated.